### PR TITLE
feat: change relation between alerts and incidents to work with fingerprints instead of alert ids

### LINKED ISSUE #2398: [➕ Feature]: Change relation between Alerts and Incidents to work with fingerprints instead of alert_ids
# Problem statement

Each alert in the database has a unique alert_id and a non-unique fingerprint, which separates a UNIQUE alert from its appearance. This allows us to track minor changes in alerts. Currently, in the UI, we always show the last appearance of the alert grouped by fingerprint.

AlertToIncident many-to-many relation right now uses alert_id as a key, and there is a runtime logic that assigns all alert_ids with the same fingerprint to the incident when we assign the last appearance. This leads to a huge size of the AlertToIncident table. 

More of that - we still need to group alerts back by fingerprint to show the last one in an incident UI.

## Current database structure
image

# Assumption

If we will refactor AlertToIncident to have a relation via fingerprint we will dramatically reduce the amount of entities in AlertToIncident tables.

Also, to simplify finding a last appearance we want to introduce a new table latest_alerts which will contain the fingerprint and last alert_id. This will allow us to avoid grouping by fingerprint to get last alert appearance

## Target database structure

image

For smooth migration, we suggest making a new table FingerprintToIncident with a new structure, then performing data migration, and then drop an old AlertToIncident table later

### PR DESCRIPTION
**Pull request by VladimirFilonov**
Closes #2398 

## ✅ Checks

- [ ] My pull request adheres to the code style of this project
- [ ] My code requires changes to the documentation
- [ ] I have updated the documentation as required
- [ ] All the tests have passed

## Release Notes

- **New Features**
  - Introduced fingerprint-based identification for alerts, replacing event IDs in incident management.
  - Added new API endpoints for fetching random workflow templates and enhanced workflow execution details.
  - Enhanced incident data with associated alerts for improved incident management.

- **Bug Fixes**
  - Improved error handling in workflow management and alert processing.

- **Documentation**
  - Updated test cases to reflect changes in alert handling and ensure accuracy in incident management.

- **Chores**
  - Enhanced test fixtures for better alert state management and performance metrics.

### CODE REVIEW

**Review by Kiryous — CHANGES_REQUESTED** [review]
Kudos, @VladimirFilonov, awesome work! Overall, looks good to me. left one minor comment

File: keep/api/core/db.py [hunk 1]
```diff
@@ -4482,3 +4487,43 @@ def get_resource_ids_by_resource_type(
...
+                    LastAlert.tenant_id == tenant_id,
+                    LastAlert.fingerprint == fingerprint,
+                )
+            )
+        ).first()
+
+def set_last_alert(
+    tenant_id: str, alert: Alert, session: Optional[Session] = None
+) -> None:
+    last_alert = get_last_alert_by_fingerprint(tenant_id, alert.fingerprint, session)
```
Comment (Kiryous) [inline keep/api/core/db.py:4509]: in theory the `alert.fingerprint` could be None. Should we handle it? 
```suggestion
    if not alert.fingerprint:
        raise AlertEmptyFingerprint()
    last_alert = get_last_alert_by_fingerprint(tenant_id, alert.fingerprint, session)
```

File: keep/api/models/db/alert.py [hunk 2]
```diff
@@ -72,6 +73,49 @@ class AlertToIncident(SQLModel, table=True):
         default=NULL_FOR_DELETED_AT,
     )
 
+class LastAlert(SQLModel, table=True):
+
+    tenant_id: str = Field(foreign_key="tenant.id", nullable=False)
+    fingerprint: str = Field(primary_key=True)
+    alert_id: UUID = Field(foreign_key="alert.id")
+    timestamp: datetime = Field(nullable=False, index=True)
```
Comment (Kiryous) [inline keep/api/models/db/alert.py:81]: Do we want to add index considering we're doing joins?

```
lastalert (fingerprint, tenant_id)
```

File: keep/api/models/db/alert.py [hunk 3]
```diff
@@ -72,6 +73,49 @@ class AlertToIncident(SQLModel, table=True):
...
+    )
+
+    is_created_by_ai: bool = Field(default=False)
+
+    deleted_at: datetime = Field(
+        default_factory=None,
+        nullable=True,
+        primary_key=True,
+        default=NULL_FOR_DELETED_AT,
+    )
```
Comment (Kiryous) [inline keep/api/models/db/alert.py:104]: also, should we consider index here? 

```
lastalerttoincident (fingerprint, tenant_id, incident_id)
```

**Review by Kiryous — APPROVED** [review]
LGTM! Left a few questions

File: keep/api/core/db.py (hunk 1 above)
Comment (VladimirFilonov) [reply keep/api/core/db.py:4557]: As far as I understand it was possible until end of September. Last entry in prod database is 2023-09-26 12:16:21.000

File: keep/api/models/db/alert.py (hunk 3 above)
Comment (VladimirFilonov) [reply keep/api/models/db/alert.py:104]: Same here

**Review by Kiryous — DISMISSED** [review]
LGTM!

File: tests/conftest.py [hunk 4]
```diff
@@ -559,8 +587,10 @@ def setup_stress_alerts(
     num_alerts = request.param.get(
         "num_alerts", 1000
     )  # Default to 1000 alerts if not specified
-
+    start_time = time.time()
     alerts = setup_stress_alerts_no_elastic(num_alerts)
+    print(f"time taken to setup {num_alerts} alerts with db: ", time.time() - start_time)
```
Comment (Matvey-Kuk) [inline tests/conftest.py:592]: logger.info.. ? :)

File: keep/api/models/db/alert.py [hunk 5]
```diff
@@ -60,8 +61,8 @@ class AlertToIncident(SQLModel, table=True):
             primary_key=True,
         )
     )
-    alert: "Alert" = Relationship(back_populates="alert_to_incident_link")
-    incident: "Incident" = Relationship(back_populates="alert_to_incident_link")
+    # alert: "Alert" = Relationship(back_populates="alert_to_incident_link")
+    # incident: "Incident" = Relationship(back_populates="alert_to_incident_link")
```
Comment (Matvey-Kuk) [inline keep/api/models/db/alert.py:65]: Leftovers?

**Review by Matvey-Kuk — CHANGES_REQUESTED** [review]

**Review by Matvey-Kuk — APPROVED** [review]

**Pull request merged**